### PR TITLE
Merge all Hubble server-side code into Cilium

### PR DESCRIPTION
**Pull request by tgraf**
This imports the server-side code of hubble to avoid depending on the
cilium/hubble repository. It does so with minimal code changes. It does
not remove unnecessary code yet.

Changes:
 - Fix all import paths
 - Replace gogo/protobuf with golang/protobuf
 - Annotate all tests with Cilium build flags
 - Replace use of sync.Mutex with lock.Mutex

### CODE REVIEW

**Review by gandro — COMMENTED** [review]
This is going to be great! A couple thoughts on package organization going forward (some of this can also be done in future PRs once this is merged):

- Where will the "standalone" Hubble server code (i.e. the `hubble serve` command plus Dockerfile for the DaemonSet) live? This PR imports all the code necessary to implement it (e.g. `pkg/hubble/ipcache`, `pkg/hubble/servicecache`), without importing the actual binary. I personally think we should keep all support code related to stand-alone mode out of the Cilium repository.

- For the Cilium API, we have the generated files in `/api/v1` and custom (non-generated) code lives in `/pkg/api/v1`. For the Hubble API, we now also have the generated files in `/api/v1`, but the custom code on `pkg/hubble/api/v1`. In my opinion, we should clean up `pkg/hubble/api/v1` (it has a ton of random definitions) and then move the remaining type definitions in to`pkg/api/v1/hubble` instead.

**Comment by Rolinh** [comment]
>     * Where will the "standalone" Hubble server code (i.e. the `hubble serve` command plus Dockerfile for the DaemonSet) live?

I think it's fine to keep it in Hubble's git repo until Cilium 1.7 is out of support (and so would hubble standalone be). The newly introduced functionalities in hubble's `cmd/serve` should be moved under `pkg/hubble/server` here imho and hubble's `pkg/server` to `pkg/hubble/observer` or so.

> In my opinion, we should clean up pkg/hubble/api/v1 (it has a ton of random definitions) and then move the remaining type definitions in topkg/api/v1/hubble instead.

:+1:

**Comment by tgraf** [comment]
> * Where will the "standalone" Hubble server code (i.e. the `hubble serve` command plus Dockerfile for the DaemonSet) live? This PR imports all the code necessary to implement it (e.g. `pkg/hubble/ipcache`, `pkg/hubble/servicecache`), without importing the actual binary. I personally think we should keep all support code related to stand-alone mode out of the Cilium repository.
> *

Agreed with @Rolinh. I've split this into a separate commit for easier review. I've just pushed the commit to remove all unused code.

> The newly introduced functionalities in hubble's `cmd/serve` should be moved under `pkg/hubble/server` here imho and hubble's `pkg/server` to `pkg/hubble/observer` or so.

I had quickly tried this but there were quite some naming conflicts and I didn't want to pollute the initial moving commit too much.

**Comment by tgraf** [comment]
> EDIT: to be clear, I think that we don't want to import this (`hubble-serve` command does not make sense in the context of embedded hubble):

I've split it into `pkg/hubble/server` and `pkg/hubble/obsever`

File: daemon/cmd/hubble.go [hunk 1]
```diff
@@ -18,17 +18,17 @@ import (
 	"strings"
 
 	"github.com/cilium/cilium/pkg/api"
-	"github.com/cilium/cilium/pkg/hubble"
+	"github.com/cilium/cilium/pkg/hubble/listener"
+	hubbleMetrics "github.com/cilium/cilium/pkg/hubble/metrics"
+	"github.com/cilium/cilium/pkg/hubble/observer"
+	"github.com/cilium/cilium/pkg/hubble/parser"
+	hubbleServer "github.com/cilium/cilium/pkg/hubble/server"
+	"github.com/cilium/cilium/pkg/hubble/server/serveroption"
```
Comment (rolinh) [inline daemon/cmd/hubble.go:26]: This should be moved under `pkg/hubble/observer`.

**Review by rolinh — CHANGES_REQUESTED** [review]

File: pkg/hubble/api/socket_test.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
+// Copyright 2019 Authors of Hubble
```
Comment (rolinh) [inline pkg/hubble/api/socket_test.go:1]: Shouldn't headers be updated to something like the following?
```
Copyright 2019 Authors of Hubble
Copyright 2020 Authors of Cilium
```

Comment (tgraf) [reply pkg/hubble/api/socket_test.go:1]: Probably, this PR doesn't really change any code so I've just left copyright notices intact for now.

**Comment by tgraf** [comment]
test-me-please

File: pkg/hubble/observer/option/option.go [hunk 3]
```diff
@@ -12,12 +12,14 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package serveroption
+package option
```
Comment (glibsm) [inline pkg/hubble/observer/option/option.go]: why this change?

**Review by michi-covalent — APPROVED** [review]

**Comment by glibsm** [comment]
Don't see a point to do a deep review here. If it compiles and it's just a code move - lets do it.

**Review by glibsm — APPROVED** [review]

**Review by rolinh — APPROVED** [review]

File: pkg/hubble/observer/option/option.go (hunk 3 above)
Comment (tgraf) [reply pkg/hubble/observer/option/option.go]: I was on requested by @Rolinh to prepare for the peer API. I'll drop the commit for now.

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Review by tklauser — APPROVED** [review]
Nice improvement on the binary size!

```
== daemon/cilium-agent ==
bss                                7813984     7812704       -1280
data                                684624      677360       -7264
text                              56297258    55761833     -535425
```
